addons/mail/models/discuss/discuss_channel.py

```diff
@@ -1087,33 +1088,66 @@ def _types_allowing_seen_infos(self):
     def channel_fetched(self):
         """ Broadcast the channel_fetched notification to channel members
         """
-        for channel in self:
-            if not channel.message_ids.ids:
-                return
-            # a bit not-modular but helps understanding code
-            if channel.channel_type not in {'chat', 'whatsapp'}:
-                return
-            last_message_id = channel.message_ids.ids[0] # zero is the index of the last message
-            member = self.env['discuss.channel.member'].search([('channel_id', '=', channel.id), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if member.fetched_message_id.id == last_message_id:
-                # last message fetched by user is already up-to-date
-                return
-            # Avoid serialization error when multiple tabs are opened.
+        to_update = []  # (member_id, last_message_id)
+        notifications = []  # channel, notification_type, message
+        partner_id = self.env.user.partner_id.id
+
+        # Avoid serialization error when multiple tabs are opened.
+        # In order to completely avoid any issues with concurrency, the very first thing we do on
+        # the cursor needs to be a SELECT FOR UPDATE
+        with self.env.registry.cursor() as cr:
+            self = self.with_env(self.env(cr=cr))
+            member_query = self.env['discuss.channel.member']._search([
+                ('channel_id', 'in', self.ids),
+                ('partner_id', '=', partner_id),
+            ])
+            all_member_result = self.env.execute_query(SQL("""
+                SELECT id, fetched_message_id, channel_id
+                  FROM discuss_channel_member
+                 WHERE id IN (%s)
+                   FOR NO KEY UPDATE SKIP LOCKED
+            """, member_query.subselect()))
+            current_member_data = {
+                channel_id: (member_id, message_id)
+                for member_id, message_id, channel_id in all_member_result
+            }
+            if not current_member_data:
+                return  # Every channel locked by another transaction
+
+            for channel in self:
+                if not channel.message_ids.ids:
+                    continue
+
+                # a bit not-modular but helps understanding code
+                if channel.channel_type not in {'chat', 'whatsapp'}:
+                    continue
+
+                if channel.id not in current_member_data:
+                    # locked by another transaction
+                    continue
+
+                last_message_id = channel.message_ids.ids[0] # zero is the index of the last message
```

Comment:
This is already there before the PR, but reading all messages from the channel just to find the last one is a really bad performance idea.

There is `_get_last_messages` but ideally the client should send as param the last message they intended to mark as fetched to avoid any confusion or race condition. And if the currently fetched is more recent, the other one should be ignored. Similar to `seen_message_id` in `_channel_seen`.

Reply:
Shouldn't the field `message_ids` be removed from the model?
Or shouldn't at least make sure to never fetch it?

Reply:
Note that `_compute_duration` is still using it in `im_livechat`.